OMX-27-firmware/storage.h

```diff
@@ -0,0 +1,71 @@
+#ifndef _INCLUDE_STORAGE_H
+#define _INCLUDE_STORAGE_H
+
+#include <Adafruit_FRAM_I2C.h>
+
+enum StorageType {
+  EEPROM_MEMORY = 0,
+  FRAM_MEMORY = 1
+};
+
+// abstract storage class
+class Storage {
+public:
+  Storage() {}
+
+  virtual int capacity() = 0;
+
+  // read/write bytes
+  virtual void write(size_t address, uint8_t val) = 0;
+  virtual uint8_t read(size_t address) = 0;
+
+// template reader/writer implementation copied from Adafruit_FRAM_I2C which implements them both
+// in terms of reading/writing bytes
+
+template <class T> uint16_t writeObject(uint16_t addr, const T &value) {
+  const byte *p = (const byte *)(const void *)&value;
+  uint16_t n;
+  for (n = 0; n < sizeof(value); n++)
+  {
+    write(addr++, *p++);
+  }
+  return n;
+}
+
+template <class T> uint16_t readObject(uint16_t addr, T &value) {
+  byte *p = (byte *)(void *)&value;
+  uint16_t n;
+  for (n = 0; n < sizeof(value); n++) {
+    *p++ = read(addr++);
+  }
+  return n;
+}
```

Comment:
These templates look fancier than they are. The compiler literally copy/pastes them with different values of `T`. the name T is a convention, you can name it whatever you want in the ` <class T>` section. In this case it just saves you from casting your objects when calling the read/write functions.

`EEPROM.get` and `EEPROM.put` use functionally identical code to this.


nitpicking here: indent these functions and include the original documentation.
```suggestion
  /**************************************************************************/
  /*!
      @brief  Write any object to memory. Originally from Adafruit_FRAM_I2C.
      @param addr
                  The 16-bit address to write to in EEPROM memory
      @param value  The templated object we will be writing
      @returns    The number of bytes written
  */
  /**************************************************************************/
  template <class T> uint16_t writeObject(uint16_t addr, const T &value) {
    const byte *p = (const byte *)(const void *)&value;
    uint16_t n;
    for (n = 0; n < sizeof(value); n++)
    {
      write(addr++, *p++);
    }
    return n;
  }

  /**************************************************************************/
  /*!
      @brief Read any object from memory.  Originally from Adafruit_FRAM_I2C.
      @param addr
                  The 16-bit address to write to in EEPROM memory
      @param value  The address of the templated object we will be writing INTO
      @returns    The number of bytes read
  */
  /**************************************************************************/
  template <class T> uint16_t readObject(uint16_t addr, T &value) {
    byte *p = (byte *)(void *)&value;
    uint16_t n;
    for (n = 0; n < sizeof(value); n++) {
      *p++ = read(addr++);
    }
    return n;
  }
```